### PR TITLE
JIT: Use reaching definitions in CSE to update conservative VNs

### LINKED ISSUE #109745: JIT: libraries jitstress Assertion failed 'doesVNMatch' during 'Assertion prop'
x86 libraries jitstress

https://dev.azure.com/dnceng-public/public/_build/results?buildId=867933&view=ms.vss-test-web.build-test-results-tab

```
"C:\h\w\A3450896\p\dotnet.exe" exec --runtimeconfig System.Collections.Immutable.Tests.runtimeconfig.json --depsfile System.Collections.Immutable.Tests.deps.json xunit.console.dll System.Collections.Immutable.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Collections.Immutable.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Collections.Immutable.Tests (found 4292 of 4397 test cases)
  Starting:    System.Collections.Immutable.Tests (parallel test collections = on [4 threads], stop on fail = off)

DOTNET_JitStress=1
DOTNET_TieredCompilation=0

Assert failure(PID 4116 [0x00001014], Thread: 9356 [0x248c]): Assertion failed 'doesVNMatch' in 'System.Collections.Immutable.Tests.ImmutableArrayTest+<ChangeType>d__192`1[int]:MoveNext():ubyte:this' during 'Assertion prop' (IL size 362; hash 0x08aa2d36; FullOpts)

    File: D:\a\_work\1\s\src\coreclr\jit\assertionprop.cpp:1486
    Image: C:\h\w\A3450896\p\dotnet.exe
```

### Known Issue Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": "Assertion failed 'doesVNMatch'",
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=867933
**Error message validated:** `[Assertion failed 'doesVNMatch'`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 11/14/2024 12:58:45 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[875331](https://dev.azure.com/dnceng-public/public/_build/results?buildId=875331)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=875331&view=ms.vss-test-web.build-test-results-tab&runId=22840146&resultId=208127)|dotnet/runtime#109987|
|[873656](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873656)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873656&view=ms.vss-test-web.build-test-results-tab&runId=22792908&resultId=208116)|dotnet/runtime#106557|
|[873420](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873420)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873420&view=ms.vss-test-web.build-test-results-tab&runId=22785798&resultId=208116)|dotnet/runtime#109826|
|[873272](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873272)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=873272&view=ms.vss-test-web.build-test-results-tab&runId=22780214&resultId=208116)|dotnet/runtime#109957|
|[872587](https://dev.azure.com/dnceng-public/public/_build/results?buildId=872587)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=872587&view=ms.vss-test-web.build-test-results-tab&runId=22757412&resultId=208095)|dotnet/runtime#109808|
|[871840](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871840)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871840&view=ms.vss-test-web.build-test-results-tab&runId=22730096&resultId=208095)|dotnet/runtime#109719|
|[871630](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871630)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871630&view=ms.vss-test-web.build-test-results-tab&runId=22723236&resultId=208096)|dotnet/runtime#109861|
|[871309](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871309)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871309&view=ms.vss-test-web.build-test-results-tab&runId=22709316&resultId=208269)|dotnet/runtime#109804|
|[871150](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871150)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871150&view=ms.vss-test-web.build-test-results-tab&runId=22705474&resultId=208095)||
|[871091](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871091)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871091&view=ms.vss-test-web.build-test-results-tab&runId=22704252&resultId=208095)|dotnet/runtime#109873|
|[871071](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871071)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=871071&view=ms.vss-test-web.build-test-results-tab&runId=22704030&resultId=208443)|dotnet/runtime#109835|
|[870154](https://dev.azure.com/dnceng-public/public/_build/results?buildId=870154)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=870154&view=ms.vss-test-web.build-test-results-tab&runId=22671266&resultId=208095)|dotnet/runtime#109503|
|[869691](https://dev.azure.com/dnceng-public/public/_build/results?buildId=869691)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=869691&view=ms.vss-test-web.build-test-results-tab&runId=22657210&resultId=208192)|dotnet/runtime#109418|
|[867933](https://dev.azure.com/dnceng-public/public/_build/results?buildId=867933)|dotnet/runtime|[System.Collections.Immutable.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=867933&view=ms.vss-test-web.build-test-results-tab&runId=22605516&resultId=211332)|dotnet/runtime#109714|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|11|14|

### PR DESCRIPTION
**Pull request by jakobbotsch**
Now that CSE always inserts into SSA we can update it to make use of the reaching definition information that it has access to. CSE already spent effort to track some extra information to try to do this, which we can remove.

- Remove `optCSECheckedBoundMap`: this was used by CSE to try to update conservative VNs of ancestor bounds checks. This is unnececssary since all descendants of the CSE definitions should get the same conservative VNs automatically now.
- Remove `CSEdsc::defConservNormVN`; this was used to update conservative VNs in the case where all defs agree on the conservative VN, which again is unnecessary now

Making this change requires a bit of refactoring to the incremental SSA builder. Before this PR the builder takes all defs and all uses and then inserts everything into SSA. After this change the builder is used in a multi-step process as follows:
1. All definitions are added with `IncrementalSsaBuilder::InsertDef`
2. The definitions are finalized with `IncrementalSsaBuilder::FinalizeDefs`
3. Uses are inserted (one by one) with `IncrementalSsaBuilder::InsertUse`. No finalization is necessary; each use is directly put into SSA as a result of calling this method.

The refactoring allows CSE to use the incremental SSA builder such that it can access reaching definition information for each of the uses as part of making replacements. However, this still requires some refactoring such that CSE performs replacements of all defs before performing the replacements of all uses.

Additionally, this PR fixes various incorrect VN updating made by CSE.

VN and CSE still track VNs that are interesting bounds check. However, VN was sometimes inserting VNs with exception sets into the set, which is not useful (the consumers always use normal VNs when querying the set). This PR fixes VN to insert the normal VN instead.

Fix #109745

### CODE REVIEW

File: src/coreclr/jit/optcse.cpp [hunk 1]
```diff
@@ -4963,325 +4832,308 @@ void CSE_HeuristicCommon::PerformCSE(CSE_Candidate* successfulCandidate)
...
+        else
+        {
+            noway_assert(store->Data() == val);
+        }
 
-            cse = cseLclVar;
-            if (isSharedConst)
+        // Assign the proper Value Numbers.
+        ValueNumPair valExc = m_pCompiler->vnStore->VNPExceptionSet(val->gtVNPair);
+        store->gtVNPair     = m_pCompiler->vnStore->VNPWithExc(ValueNumStore::VNPForVoid(), valExc);
```
Comment (jakobbotsch) [inline src/coreclr/jit/optcse.cpp:4891]: This was assigning `ValueNumStore::VNPForVoid` before, without any exceptions. That seems wrong.

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -12719,7 +12719,8 @@ void Compiler::fgValueNumberTree(GenTree* tree)
 
                         // Record non-constant value numbers that are used as the length argument to bounds checks, so
                         // that assertion prop will know that comparisons against them are worth analyzing.
-                        ValueNum lengthVN = tree->AsBoundsChk()->GetArrayLength()->gtVNPair.GetConservative();
+                        ValueNum lengthVN =
+                            vnStore->VNNormalValue(tree->AsBoundsChk()->GetArrayLength()->gtVNPair.GetConservative());
```
Comment (jakobbotsch) [inline src/coreclr/jit/valuenum.cpp:12723]: All consumers of this info query the normal conservative VN, so inserting one with exceptions is uninteresting. I hit a few diffs related to this.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=874716&view=ms.vss-build-web.run-extensions-tab). Some minor CQ and TP improvements.

**Review by AndyAyersMS — APPROVED** [review]
Did you ever collect stats on how often we bail out putting defs into SSA because of IDF size?

**Comment by jakobbotsch** [comment]
> Did you ever collect stats on how often we bail out putting defs into SSA because of IDF size?

Here's some histograms for IDF size we see in the incremental builder for win-x64 collections:

```scala
aspnet
IDF size
     <=          1 ===>     779 count ( 15% of total)
      2 ..       2 ===>     780 count ( 30% of total)
      3 ..       3 ===>     639 count ( 43% of total)
      4 ..       4 ===>     608 count ( 55% of total)
      5 ..       5 ===>     380 count ( 63% of total)
      6 ..      10 ===>    1366 count ( 90% of total)
     11 ..      20 ===>     386 count ( 97% of total)
     21 ..      30 ===>      95 count ( 99% of total)
     31 ..      40 ===>       9 count ( 99% of total)
     41 ..      50 ===>       4 count (100% of total)
     51 ..     100 ===>       0 count (100% of total)
    101 ..     200 ===>       0 count (100% of total)
    201 ..     300 ===>       0 count (100% of total)
    301 ..     400 ===>       0 count (100% of total)
    401 ..     500 ===>       0 count (100% of total)

benchmarks.run_pgo
IDF size
     <=          1 ===>     298 count ( 12% of total)
      2 ..       2 ===>     531 count ( 35% of total)
      3 ..       3 ===>     200 count ( 43% of total)
      4 ..       4 ===>     243 count ( 54% of total)
      5 ..       5 ===>     170 count ( 61% of total)
      6 ..      10 ===>     770 count ( 94% of total)
     11 ..      20 ===>     114 count ( 99% of total)
     21 ..      30 ===>       8 count ( 99% of total)
     31 ..      40 ===>       0 count ( 99% of total)
     41 ..      50 ===>       7 count ( 99% of total)
     51 ..     100 ===>       1 count (100% of total)
    101 ..     200 ===>       0 count (100% of total)
    201 ..     300 ===>       0 count (100% of total)
    301 ..     400 ===>       0 count (100% of total)
    401 ..     500 ===>       0 count (100% of total)


libraries_tests.run:
IDF size
     <=          1 ===>    4340 count ( 18% of total)
      2 ..       2 ===>    3915 count ( 35% of total)
      3 ..       3 ===>    3542 count ( 50% of total)
      4 ..       4 ===>    3435 count ( 64% of total)
      5 ..       5 ===>    1806 count ( 72% of total)
      6 ..      10 ===>    5126 count ( 94% of total)
     11 ..      20 ===>    1231 count ( 99% of total)
     21 ..      30 ===>     151 count ( 99% of total)
     31 ..      40 ===>      20 count ( 99% of total)
     41 ..      50 ===>       5 count ( 99% of total)
     51 ..     100 ===>       2 count (100% of total)
    101 ..     200 ===>       0 count (100% of total)
    201 ..     300 ===>       0 count (100% of total)
    301 ..     400 ===>       0 count (100% of total)
    401 ..     500 ===>       0 count (100% of total)

realworld:
IDF size
     <=          1 ===>     449 count ( 22% of total)
      2 ..       2 ===>     366 count ( 41% of total)
      3 ..       3 ===>     341 count ( 58% of total)
      4 ..       4 ===>     225 count ( 70% of total)
      5 ..       5 ===>     178 count ( 79% of total)
      6 ..      10 ===>     305 count ( 95% of total)
     11 ..      20 ===>      81 count ( 99% of total)
     21 ..      30 ===>      13 count ( 99% of total)
     31 ..      40 ===>       1 count ( 99% of total)
     41 ..      50 ===>       1 count (100% of total)
     51 ..     100 ===>       0 count (100% of total)
    101 ..     200 ===>       0 count (100% of total)
    201 ..     300 ===>       0 count (100% of total)
    301 ..     400 ===>       0 count (100% of total)
    401 ..     500 ===>       0 count (100% of total)
```

So looks like none of these collections have any examples that fail SSA insertion.

**Comment by jakobbotsch** [comment]
Some examples of functions with large (>= 40) IDF sizes for some CSEs.

https://github.com/PowerShell/PowerShell/blob/7ca7aae1d13d19e38c7c26260758f474cb9bef7f/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L1513-L3580 (a 2000 line function)

https://github.com/dotnet/performance/blob/4894b54b3e86637b040ba6c7c54c23524cfbeadd/src/benchmarks/micro/runtime/Bytemark/assign_rect.cs#L438-L540 (the OSR version of this)
The CSE here is for `tableau.GetLength(1)`.

https://github.com/dotnet/runtime/blob/6717d71fca1fac5f70fd2f8cda107d8c4320ab6a/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs#L835-L1522 (standard tier 1 version).
This one has a CSE with 27 definitions and with no uses in reachable blocks. I wonder if it would be worth it to compute the DFS tree. It seems like the heuristic is not accounting for reachability of the uses (probably the better solution is to aggressively trim the blocks earlier)

**Pull request merged**